### PR TITLE
Fix rope equality hacks

### PR DESCRIPTION
**Pull request by nirvdrum**

### CODE REVIEW

File: src/main/java/org/truffleruby/core/string/FrozenStrings.java [hunk 1]
```diff
@@ -51,12 +50,12 @@ public DynamicObject getFrozenString(Rope rope) {
 
         lock.writeLock().lock();
         try {
-            string = frozenStrings.get(holder);
+            string = frozenStrings.get(rope);
             if (string == null) {
                 string = StringOperations.createFrozenString(context, rope);
                 // TODO CS 8-Jan-18 as in RopeCache the map is weak, but the key is a wrapper! So we just hold on to everything forever using this method.
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/core/string/FrozenStrings.java]: This comment doesn't apply any more. The key is no longer a wrapper.

The key set here should probably remain as `RopeCache` is still not correct, but the comment needs to be updated.

I think the underlying problem in both `RopeCache` and `FrozenStrings` is that it needs to be a weak-valued hash map, as well as a weak-valued one. We're doing that correctly elsewhere.

In this PR we just need to fix the comment for now though.

File: src/main/java/org/truffleruby/parser/lexer/RubyLexer.java [hunk 2]
```diff
@@ -3084,7 +3084,7 @@ public void setValue(Object yaccValue) {
     }
 
     protected boolean strncmp(Rope one, Rope two, int length) {
-        if (one.byteLength() < length || two.byteLength() < length) {
+        if (one.byteLength() != two.byteLength() && (one.byteLength() < length || two.byteLength() < length)) {
```
Comment (eregon) [inline src/main/java/org/truffleruby/parser/lexer/RubyLexer.java]: Does it make sense to call `strncmp` when length is bigger than one of the two Ropes?
Or should this be some kind of `assert`?

What should happen if `one.byteLength() == two.byteLength()` but `length > one.byteLength()`?

File: src/main/java/org/truffleruby/core/string/FrozenStrings.java [hunk 3]
```diff
@@ -24,10 +24,10 @@
 
     private final RubyContext context;
 
-    private final Map<RopeHolder, DynamicObject> frozenStrings = new WeakHashMap<>();
+    private final Map<Rope, DynamicObject> frozenStrings = new WeakHashMap<>();
     private final ReadWriteLock lock = new ReentrantReadWriteLock();
 
-    private final Set<RopeHolder> keys = new HashSet<>();
+    private final Set<Rope> keys = new HashSet<>();
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/string/FrozenStrings.java]: Does this commit enables to remove this `keys` set?
Maybe for another PR?

Comment (chrisseaton) [reply src/main/java/org/truffleruby/core/string/FrozenStrings.java]: See my comment below. I will tackle this in another PR when this is merged.

File: src/main/java/org/truffleruby/parser/lexer/RubyLexer.java (hunk 2 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/parser/lexer/RubyLexer.java]: I'm going to look at cleaning this up in another PR. I should've just left the short-circuit check alone and can back that commit out if you'd prefer. But it struck me as being blatantly wrong.

Comment (eregon) [reply src/main/java/org/truffleruby/parser/lexer/RubyLexer.java]: I think in the case I described above, the `assert` in `ArrayUtils.memcmp` is going to trigger.
And I feel there is a semantic problem here, we should only `strncmp` if length is smaller than both Ropes' byteLengths.
Feel free to fix in another PR though.

File: src/main/java/org/truffleruby/core/string/FrozenStrings.java (hunk 3 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/core/string/FrozenStrings.java]: I think removing the `keys` is correct, but I wasn't entirely sure and was going to let @chrisseaton handle that given he just did some major work to `FrozenStrings`.

File: src/main/java/org/truffleruby/parser/lexer/RubyLexer.java (hunk 2 above)
Comment (nirvdrum) [reply src/main/java/org/truffleruby/parser/lexer/RubyLexer.java]: It's code we inherited and it seems to work. The length is chosen as the length of one of the ropes in all usages, so at least one of the ropes is never smaller than length. But, yes, let's defer to another PR.

**Pull request merged**